File: latte/src/main/java/gg/beemo/latte/broker/BrokerMessageHeaders.kt
package gg.beemo.latte.broker

import java.util.UUID

open class BrokerMessageHeaders(val headers: Map<String, String>) {

    val sourceService: String by lazy {
        headers.getOrThrow(HEADER_SOURCE_SERVICE)
    }
    val sourceInstance: String by lazy {
        headers.getOrThrow(HEADER_SOURCE_INSTANCE)
    }
    val targetServices: Set<String> by lazy {
        splitToSet(headers.getOrDefault(HEADER_TARGET_SERVICES, ""))
    }
    val targetInstances: Set<String> by lazy {
        splitToSet(headers.getOrDefault(HEADER_TARGET_INSTANCES, ""))
    }
    val messageId: MessageId by lazy {
        headers.getOrThrow(HEADER_MESSAGE_ID)
    }

    constructor(
        connection: BrokerConnection,
        targetServices: Set<String>,
        targetInstances: Set<String>,
    ) : this(
        createHeadersMap(
            connection.serviceName,
            connection.instanceId,
            targetServices,
            targetInstances,
            null,
        ),
    )

    companion object {

        private const val HEADER_SOURCE_SERVICE = "source-service"
        private const val HEADER_SOURCE_INSTANCE = "source-instance"
        private const val HEADER_TARGET_SERVICES = "target-services"
        private const val HEADER_TARGET_INSTANCES = "target-instances"
        private const val HEADER_MESSAGE_ID = "message-id"

        // Needs to be JvmStatic to be used in subclasses
        @JvmStatic
        protected fun createHeadersMap(
            sourceService: String,
            sourceInstance: String,
            targetServices: Set<String>,
            targetInstances: Set<String>,
            messageId: MessageId?,
            extra: Map<String, String> = emptyMap(),
        ): Map<String, String> {
            val headers = HashMap<String, String>()
            headers[HEADER_SOURCE_SERVICE] = sourceService
            headers[HEADER_SOURCE_INSTANCE] = sourceInstance
            headers[HEADER_TARGET_SERVICES] = targetServices.joinToString(",")
            headers[HEADER_TARGET_INSTANCES] = targetInstances.joinToString(",")
            headers[HEADER_MESSAGE_ID] = messageId ?: UUID.randomUUID().toString()
            headers.putAll(extra)
            return headers
        }

        protected fun splitToSet(value: String): Set<String> {
            return value.split(",").filter { it.isNotEmpty() }.toSet()
        }

    }

}

internal fun Map<String, String>.getOrThrow(key: String): String {
    return get(key) ?: throw IllegalArgumentException("Missing broker message header '$key'")
}

File: water/src/broker/BrokerMessageHeaders.ts
import { randomUUID } from "node:crypto";
import { BrokerConnection } from "../index.js";
import type { MessageId } from "./BrokerConnection.js";

const BROKER_HEADER_NAMES = {
    SOURCE_SERVICE: "source-service",
    SOURCE_INSTANCE: "source-instance",
    TARGET_SERVICES: "target-services",
    TARGET_INSTANCES: "target-instances",
    MESSAGE_ID: "message-id",
};

export class BrokerMessageHeaders {

    public readonly headers: Record<string, string>;

    public get sourceService(): string {
        return getOrThrow(this.headers, BROKER_HEADER_NAMES.SOURCE_SERVICE);
    }
    public get sourceInstance(): string {
        return getOrThrow(this.headers, BROKER_HEADER_NAMES.SOURCE_INSTANCE);
    }
    public get targetServices(): Set<string> {
        return new Set((this.headers[BROKER_HEADER_NAMES.TARGET_SERVICES] ?? "").split(",").filter(s => s));
    }
    public get targetInstances(): Set<string> {
        return new Set((this.headers[BROKER_HEADER_NAMES.TARGET_INSTANCES] ?? "").split(",").filter(s => s));
    }
    public get messageId(): MessageId {
        return getOrThrow(this.headers, BROKER_HEADER_NAMES.MESSAGE_ID);
    }

    public constructor(headers: Record<string, string>);
    public constructor(
        connection: BrokerConnection,
        targetServices: Set<string>,
        targetInstances: Set<string>,
    );
    public constructor(
        headersOrConnection: BrokerConnection | Record<string, string>,
        targetServices?: Set<string>,
        targetInstances?: Set<string>,
    ) {
        // I hate that JS doesn't have native method overloading
        if (headersOrConnection instanceof BrokerConnection) {
            const connection = headersOrConnection;
            this.headers = BrokerMessageHeaders.createHeadersMap(
                connection.serviceName,
                connection.instanceId,
                targetServices!,
                targetInstances!,
                null,
            );
        } else {
            this.headers = headersOrConnection;
        }
    }

    protected static createHeadersMap(
        sourceService: string,
        sourceInstance: string,
        targetServices: Set<string>,
        targetInstances: Set<string>,
        messageId: MessageId | null,
        extra: Record<string, string> = {},
    ): Record<string, string> {
        const headers: Record<string, string> = {
            [BROKER_HEADER_NAMES.SOURCE_SERVICE]: sourceService,
            [BROKER_HEADER_NAMES.SOURCE_INSTANCE]: sourceInstance,
            [BROKER_HEADER_NAMES.TARGET_SERVICES]: [...targetServices].join(","),
            [BROKER_HEADER_NAMES.TARGET_INSTANCES]: [...targetInstances].join(","),
            [BROKER_HEADER_NAMES.MESSAGE_ID]: messageId ?? randomUUID(),
            ...extra,
        };
        return headers;
    }

}

export function getOrThrow(headers: Record<string, string>, key: string): string {
    const value = headers[key];
    if (value === undefined) {
        throw new Error(`Missing broker message header '${key}'`);
    }
    return value;
}
